### DISCUSSION TITLE
File Linking API

### DISCUSSION THREAD

**Discussion by tmorrell** [inveniosoftware/product-rdm#169] [Ideas]
## Is your feature request related to a problem? Please describe.

- We have files on an external storage system that we don't want to have to upload a second time

- We have files on an external system that are constantly changing

- Uploads through InvenioRDM are slow

- I want to use a custom upload tool to do my file uploads

## Describe the solution you'd like
A new API to link files to an existing RDM record. This is designed for use by administrators who've decided that a file link is acceptable.

This will accept the url of the file, file name, size (optional), and checksum (optional) and put the information into the files structure in RDM. It will then display the file download links in the UI (same file section as normal files probably).

Feature is configurable in invenio.cfg

RDM background processes will be able to conduct checksum verification and link verification

## Describe alternatives you've considered
A multipart uploader implementation can solve the slow upload parts of this issue

## Additional context
From RDM Partner Meeting

**Comment by tmorrell** [comment]
We've got a minimally working prototype. Related PR: https://github.com/inveniosoftware/invenio-records-resources/pull/567

Work remaining

- [ ] Finish implementation of remote transfer (checksums, size)
- [ ] FILES_REST_STORAGE_FACTORY should be subclassed records-resources and then import into app-rdm. storage This does send_file redirect for files. new remote_file_storage class with a send_file method that does the redirect.
- [ ] Permissions currently in rdm_records for viewing files needs to be modified.
- [ ] Adjust domain configuration. Connected to the transfer type work (RECORDS_RESOURCES_FILES_ALLOWED_DOMAINS is too general and should be defined per transfer type)
- [ ] Make sure that usage statistics are properly tracked for remote files, i.e. that the `[/api]/records/{id}/files/{key}/content` URL still generates a download event.

**Comment by Herrner** [comment]
relates to #60

**Comment by ntarocco** [comment]
@slint working on this

**Comment by alahiff** [comment]
Hi, I was wondering what the status of this is since it hasn't been updated for a while? Thanks!

**Reply by slint** [reply]
This is now being addressed by two connected initiatives:

- Allow customizable file transfer backends (including support for the `Remote` type for external file linking): https://github.com/inveniosoftware/rfcs/pull/91
- A new file upload UI, that would allow users to supply remote URLs: https://github.com/inveniosoftware/rfcs/issues/86

I know @tmorrell is also involved in these, so I assume we're on a good path. We expect this to be also discussed further in the [2025 InvenioRDM Partners Workshop in Hamburg](https://herrner.github.io/irdm2025/).

**Comment by ntarocco** [comment]
Implemented and available in InvenioRDM v13
